#Проект по SQL

**Описание проекта:**
Компания купила крупный сервис для чтения книг по подписке во время короновируса. Надо проанализировать данные о книгах, издательствах и авторах.


**Цель проекта:**
Сформулировать ценностное предложение для нового продукта.

**Описание данных:**

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

## Знакомство с таблицами

In [ ]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp',  # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
db_config['pwd'],
db_config['host'],
db_config['port'],
db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

##Прочитаем таблицы

In [ ]:
query = """
SELECT *
FROM books
LIMIT 5;
"""
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [ ]:
query = """
SELECT *
FROM authors
LIMIT 5;
"""
pd.io.sql.read_sql(query, con = engine)


author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [ ]:
query = """
SELECT *
FROM publishers
LIMIT 5;
"""
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [ ]:
query = """
SELECT *
FROM ratings
LIMIT 5;
"""
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [ ]:
query = """
SELECT *
FROM reviews
LIMIT 5;
"""
pd.io.sql.read_sql(query, con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

##Перейдем к заданиям

### Задание 1. Посчитайте, сколько книг вышло после 1 января 2000 года

In [ ]:
query = """
SELECT COUNT(book_id)
FROM books
WHERE publication_date > '2000-01-01'
"""
pd.io.sql.read_sql(query, con = engine)


count
0    819

Вывод: после 1 января 2000 года вышло 819 книг

### Задание 2. Для каждой книги посчитайте количество обзоров и среднюю оценку

In [ ]:
query = """
WITH rating_all AS
    (SELECT AVG(ra.rating) AS avg_score,
            b.book_id,
            b.title
     FROM books as b
     LEFT JOIN ratings AS ra ON ra.book_id = b.book_id 
     GROUP BY b.book_id),
     
review_all AS
    (SELECT COUNT(review_id) AS reviews_count,
            b.book_id,
            b.title
     FROM books as b
     LEFT JOIN reviews AS re ON re.book_id = b.book_id 
     GROUP BY b.book_id)

SELECT b.title,
       b.book_id,
       avg_score,
       reviews_count
FROM books as b
       INNER JOIN rating_all as ra_1 ON b.book_id = ra_1.book_id
       INNER JOIN review_all as re_1 ON b.book_id = re_1.book_id
ORDER BY reviews_count DESC
"""

pd.io.sql.read_sql(query, con = engine)

title  book_id  avg_score  \
0                              Twilight (Twilight  #1)      948   3.662500   
1    The Curious Incident of the Dog in the Night-Time      695   4.081081   
2    Harry Potter and the Chamber of Secrets (Harry...      299   4.287500   
3               The Da Vinci Code (Robert Langdon  #2)      696   3.830508   
4                                     The Glass Castle      734   4.206897   
..                                                 ...      ...        ...   
995   The Cat in the Hat and Other Dr. Seuss Favorites      672   5.000000   
996                            The Natural Way to Draw      808   3.000000   
997                          Essential Tales and Poems      221   4.000000   
998    Anne Rice's The Vampire Lestat: A Graphic Novel       83   3.666667   
999                               Leonardo's Notebooks      387   4.000000   

     reviews_count  
0                7  
1                6  
2                6  
3                6  
4                6  
..             ...  
995              0  
996              0  
997              0  
998              0  
999              0  

[1000 rows x 4 columns]

Вывод: Книга с самым большим рейтингом - Twilight.

### Задание 3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [ ]:
query = """
SELECT publisher,
       COUNT(b.book_id) as books_amount
FROM books as b
     LEFT JOIN publishers as p ON b.publisher_id = p.publisher_id
WHERE num_pages > 50
GROUP BY publisher
ORDER BY books_amount DESC
LIMIT 1;
"""
pd.io.sql.read_sql(query, con = engine)

publisher  books_amount
0  Penguin Books            42

Вывод: Издательство, которое выпустило наибольше количество книг (42), более 50 страниц, - Penguin Books. 

###Задание 4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [ ]:
query = """
SELECT author,
       AVG(rating) as avg_rating
FROM books as b 
     LEFT JOIN ratings as ra ON b.book_id = ra.book_id
     LEFT JOIN authors as a ON b.author_id = a.author_id
WHERE b.book_id IN (
                   SELECT book_id
                   FROM ratings 
                   GROUP BY book_id
                   HAVING COUNT(rating_id) >= 50
                   ORDER BY COUNT(rating_id) DESC)
GROUP BY author
ORDER BY avg_rating DESC
LIMIT 1;
"""
pd.io.sql.read_sql(query, con = engine)

author  avg_rating
0  J.K. Rowling/Mary GrandPré    4.287097

Вывод: Автор книги, более 50 страниц, с самым большим среднем рейтингом - J.K. Rowling/Mary GrandPré. Это не удивительно - Гарри Потера любят многие)

###Задание 5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [ ]:
query = """
SELECT ROUND(AVG(review_count))
FROM (SELECT COUNT(review_id) AS review_count
       FROM reviews 
       WHERE username IN (
                          SELECT username
                          FROM ratings
                          GROUP BY username
                          HAVING COUNT(rating_id) > 50)
       GROUP BY username) AS a
"""
pd.io.sql.read_sql(query, con = engine)

round
0   24.0

Вывод: Среднее количество обзоров среди пользователей, которые поставили более 50 оценок, равно 24.

**Итоговый вывод:**
1. После 01.01.2000 вышло 819 книг - довольно небольшое количество за фактически 20 лет
2. Книга с самым большим рейтингом - A Dirty Job. В начале 2000-х это и правда был бестселлер по версии New York Times
3. Penguin Books выпустило 42 книги, более 50 страниц, - это самый большой показатель среди издательств
4. Джоан Роулинг - автор с самым большим рейтингом среди книг, объемом более 50 книг
5. Среднее количество обзоров среди пользователей, которые поставили более 50 оценок, равно 24 - довольно большой показатель
